Adds a method that returns the Kelly bet for a market

JonathanNankivell opened the pull request:
The standard formula for the Kelly bet assumes that the odds are fixed. This is equivalent to assuming infinite liquidity - hardly reasonable when it comes to Manifold Markets.

Naively using the standard Kelly formula may lead to ruin and despair as it overestimates the number of shares a bet will buy. 

It's not hard, however, to calculate the number of shares we will actually buy. It is then easy to find the true Kelly bet.

JonathanNankivell committed:
Implement Kelly Calculator

JonathanNankivell committed:
Updated README.md to include info about Kelly bets

bcongdon commented on pymanifold/lib.py:
```diff
@@ -149,3 +151,87 @@ def _create_market(
         )
         return LiteMarket.from_dict(response.json())
+    
+    def kelly_calc(self, slug, subjective_probability, balance, initial_market_probability=0.5):
```
I think it might make more sense to have this be in a `utils` package (e.g. `pymanifold.utils.kelly_calc`), and have the function accept a market -- since this function is more a calculation on a market, rather than an action needing to perform an API action

e.g. `kelly_calc(market, subjective_probability, balance, initial_market_probability)`.

What do you think?

JonathanNankivell replied:
I'm a maths student and might be unfamiliar with software engineering standards. I'm not sure how to create a `utils ` package (or why that would be advisable). Could you please elaborate?

Your second point is well taken and shouldn't be too hard.

JonathanNankivell replied:
I have changed the `kelly_calc` method to accept a market object rather than a slug.

JonathanNankivell committed:
Change `kelly_calc` to accept a market rather than a slug

JonathanNankivell committed:
Updated example since `kelly_calc` takes a market

JonathanNankivell committed:
Automatically find the initial market probability

Added a snippet of code that finds the probability before the first be was made. It is this probability that gets used in all the later Maniswap calculations.

JonathanNankivell committed:
Update TODO

JonathanNankivell committed:
Cleaned code

JonathanNankivell committed:
Remove `ManifoldClient.kelly_calc`

JonathanNankivell committed:
Create utils.py

Added `kelly_calc` to `utils.py`. Hopefully that is what creating a `utils` package means?

bcongdon commented on README.md:
```diff
@@ -31,13 +31,25 @@ market = client.create_binary_market(
+
+# Find optimal Kelly bet
+market = client.get_market_by_slug("for-this-study-published-in-nature")
+client.kelly_calc(
```
Can you update this example to reflect the new location of the function?

bcongdon commented on README.md:
```diff
@@ -31,13 +31,25 @@ market = client.create_binary_market(
 - [ ] Publish a version of this package to PyPI
 - [ ] Add instructions for running tests that require an API key (e.g. setting the `MANIFOLD_API_KEY` environment variable)
 
+- [ ] Generalise to correlated markets
```
nit: "Generalise" -> "Generalize `kelly_calc`"

bcongdon commented on pymanifold/lib.py:
```diff
@@ -1,4 +1,6 @@
 import requests
+from numpy import log as ln, argmax
```
Can you add numpy to the project's requirements? https://github.com/bcongdon/PyManifold/blob/main/pyproject.toml

JonathanNankivell committed:
Updated example & fixed UK spelling

bcongdon reviewed (approved):
Thanks for adding this!
